custom_components/tahoma/coordinator.py

```diff
@@ -43,20 +47,29 @@ def __init__(
         self.listener_id = listener_id
 
     async def _async_update_data(self) -> Dict[str, Device]:
-        """Fetch data from Tahoma."""
+        """Fetch TaHoma data via event listener."""
         try:
             events = await self.client.fetch_event_listener(self.listener_id)
-        except Exception as exception:
-            raise UpdateFailed(f"Error communicating with the TaHoma API: {exception}")
-        else:
-            for event in events:
-                if event.name == "DeviceStateChangedEvent":
-                    for state in event.device_states:
-                        device = self.devices[event.deviceurl]
-                        if state.name not in device.states:
-                            device.states[state.name] = state
-                        device.states[state.name].value = self._get_state(state)
+        except (ServerDisconnectedError, NotAuthenticatedException) as exception:
+            _LOGGER.debug(exception)
+            await self.client.login()
+            self.listener_id = await self.client.register_event_listener()
+            self.devices = {
+                d.deviceurl: d for d in await self.client.get_devices(refresh=True)
+            }
             return self.devices
+        except Exception as exception:
+            raise UpdateFailed(exception)
+
+        for event in events:
```

Comment:
events can be None in case we got NotAuthenticatedException.

Reply:
I do `return self.devices` above, thus this code is never reached in that scenario.

Reply:
Oh indeed sorry, the blank space trick me.